src/vivarium/framework/engine.py

```diff
@@ -35,107 +33,139 @@
+        self._component_manager.add_managers(managers)
+
+        component_config_parser = self._plugin_manager.get_plugin('component_configuration_parser')
+        # Tack extra components onto the end of the list generated from the model specification.
+        components = (component_config_parser.get_components(self._component_configuration)
+                      + self._additional_components
+                      + [Metrics()])
+        self.add_components(components)
 
     def setup(self):
-        self.lifecycle.set_state('setup')
-        self.component_manager.setup_components(self.builder)
-        self.simulant_creator = self.builder.population.get_simulant_creator()
+        self._lifecycle.set_state('setup')
+        self._component_manager.setup_components(self._builder)
 
-        self.time_step_events = self.lifecycle.get_states('main_loop')
-        self.time_step_emitters = {k: self.builder.event.get_emitter(k) for k in self.time_step_events}
-        self.end_emitter = self.builder.event.get_emitter('simulation_end')
-        self.configuration.freeze()
+        self.simulant_creator = self._builder.population.get_simulant_creator()
 
-        # TODO: Remove when lifecycle restrictions are implemented.
-        self._setup = True
+        self.time_step_events = self._lifecycle.get_states('main_loop')
+        self.time_step_emitters = {k: self._builder.event.get_emitter(k) for k in self.time_step_events}
+        self.end_emitter = self._builder.event.get_emitter('simulation_end')
+        self.configuration.freeze()
 
-        self.lifecycle.set_state('post_setup')
-        self.builder.event.get_emitter('post_setup')(None)
+        self._lifecycle.set_state('post_setup')
+        self._builder.event.get_emitter('post_setup')(None)
 
     def initialize_simulants(self):
-        self.lifecycle.set_state('population_creation')
+        self._lifecycle.set_state('population_creation')
 
         pop_params = self.configuration.population
         # Fencepost the creation of the initial population.
-        self.clock.step_backward()
+        self._clock.step_backward()
         population_size = pop_params.population_size
         self.simulant_creator(population_size, population_configuration={'sim_state': 'setup'})
-        self.clock.step_forward()
+        self._clock.step_forward()
 
     def step(self):
-        _log.debug(self.clock.time)
+        _log.debug(self._clock.time)
         for event in self.time_step_events:
-            self.lifecycle.set_state(event)
-            self.time_step_emitters[event](self.population.get_population(True).index)
-        self.clock.step_forward()
+            self._lifecycle.set_state(event)
+            self.time_step_emitters[event](self._population.get_population(True).index)
+        self._clock.step_forward()
+
+    def run(self):
+        while self._clock.time < self._clock.stop_time:
```

Comment:
Run has been added (it was already in the interactive context), finalize and report have been slightly expanded to incorporate some of the things that were in the pure functions in this module.